server/functions/framework/sql_function.go

```diff
@@ -0,0 +1,190 @@
+	return sqlFunc.Strict
+}
+
+// NonDeterministic implements the interface FunctionInterface.
+func (sqlFunc SQLFunction) NonDeterministic() bool {
+	return sqlFunc.IsNonDeterministic
+}
+
+// VariadicIndex implements the interface FunctionInterface.
+func (sqlFunc SQLFunction) VariadicIndex() int {
+	// TODO: implement variadic
+	return -1
+}
+
+// IsSRF implements the interface FunctionInterface.
+func (sqlFunc SQLFunction) IsSRF() bool {
+	return false
+}
+
+// enforceInterfaceInheritance implements the interface FunctionInterface.
+func (sqlFunc SQLFunction) enforceInterfaceInheritance(error) {}
+
+// CallSqlFunction runs the given SQL definition inside the function on the given runner.
+func CallSqlFunction(ctx *sql.Context, f SQLFunction, runner sql.StatementRunner, args []any) (any, error) {
+	paramMap := make(map[string]string)
+	for i, name := range f.ParameterNames {
+		formattedVar, err := f.ParameterTypes[i].FormatValue(args[i])
+		if err != nil {
+			return nil, err
+		}
+		if name == "" {
+			// sanity check
+			name = fmt.Sprintf("$%d", i+1)
+		}
+		paramMap[name] = formattedVar
+	}
+
+	query, err := parseAndReplaceFunctionColumn(ctx, f.SqlStatement, paramMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return sql.RunInterpreted(ctx, func(subCtx *sql.Context) (any, error) {
+		sch, rowIter, _, err := runner.QueryWithBindings(ctx, query, nil, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		if !f.SetOf {
+			// single row result
+			rows, err := sql.RowIterToRows(subCtx, rowIter)
+			if err != nil {
+				return nil, err
+			}
+			if len(sch) != 1 {
+				return nil, errors.New("expression does not result in a single value")
+			}
+			if len(rows) != 1 {
+				return nil, errors.New("expression returned multiple result sets")
+			}
+			if len(rows[0]) != 1 {
+				return nil, errors.New("expression returned multiple results")
+			}
+			return rows[0][0], nil
+		}
+		// multiple row result
+		return rowIter, nil
+	})
+}
+
+// parseAndReplaceFunctionColumn parses and replaces function parameter expressions with given arguments.
+func parseAndReplaceFunctionColumn(ctx *sql.Context, q string, params map[string]string) (string, error) {
+	parsed, err := parser.ParseOne(q)
+	if err != nil {
+		return "", err
+	}
+
+	// Function's final statement must be SELECT or INSERT/UPDATE/DELETE RETURNING
+	switch s := parsed.AST.(type) {
+	case *tree.Select:
+		sc := s.Select.(*tree.SelectClause)
+		for i, e := range sc.Exprs {
+			sc.Exprs[i].Expr = replaceToFunctionColumn(params, e.Expr)
+		}
+		if sc.Where != nil {
+			sc.Where.Expr = replaceToFunctionColumn(params, sc.Where.Expr)
+		}
+	}
+
+	return parsed.AST.String(), nil
+}
+
+// replaceToFunctionColumn replaces Placeholder and UnresolvedName expressions with FunctionColumn containing
+// argument value if applicable when the name of expression matches function parameter.
+func replaceToFunctionColumn(paramMap map[string]string, expr tree.Expr) tree.Expr {
```

Comment:
This looks exactly the same as in `server/ast/create_function.go`, we could probably export one of them and reference it from the other package.